Sources/SwiftLayout/Refactoring/Protocol/LayoutTraversal.swift

```diff
@@ -0,0 +1,201 @@
+        switch layout {
+        case let .trueLayout(layout):
+            cast(layout)?.traverse(superview, continueAfterViewLayout: continueAfterViewLayout, traverseHandler: handler)
+        case let .falseLayout(layout):
+            cast(layout)?.traverse(superview, continueAfterViewLayout: continueAfterViewLayout, traverseHandler: handler)
+        }
+    }
+    func traverse(_ superview: UIView?, viewInfoSet: ViewInformationSet, constraintHndler handler: (UIView?, UIView, [Constraint], ViewInformationSet) -> Void) {
+        switch layout {
+        case let .trueLayout(layout):
+            cast(layout)?.traverse(superview, viewInfoSet: viewInfoSet, constraintHndler: handler)
+        case let .falseLayout(layout):
+            cast(layout)?.traverse(superview, viewInfoSet: viewInfoSet, constraintHndler: handler)
+        }
+    }
+}
+
+extension OptionalLayout: LayoutTraversal {
+    func traverse(_ superview: UIView?, continueAfterViewLayout: Bool, traverseHandler handler: TraverseHandler) {
+        switch layout {
+        case let .some(layout):
+            cast(layout)?.traverse(superview, continueAfterViewLayout: continueAfterViewLayout, traverseHandler: handler)
+        case .none:
+            break
+        }
+    }
+    func traverse(_ superview: UIView?, viewInfoSet: ViewInformationSet, constraintHndler handler: (UIView?, UIView, [Constraint], ViewInformationSet) -> Void) {
+        switch layout {
+        case let .some(layout):
+            cast(layout)?.traverse(superview, viewInfoSet: viewInfoSet, constraintHndler: handler)
+        case .none:
+            break
+        }
+    }
+}
+
+extension SublayoutLayout: LayoutTraversal {
+    func traverse(_ superview: UIView?, continueAfterViewLayout: Bool, traverseHandler handler: TraverseHandler) {
+        cast(superlayout)?.traverse(superview, continueAfterViewLayout: continueAfterViewLayout, traverseHandler: { superview, subview, identifier, animationDisabled in
+            handler(superview, subview, identifier, animationDisabled)
+            cast(sublayout)?.traverse(subview, continueAfterViewLayout: continueAfterViewLayout, traverseHandler: handler)
+        })
+    }
+    func traverse(_ superview: UIView?, viewInfoSet: ViewInformationSet, constraintHndler handler: (UIView?, UIView, [Constraint], ViewInformationSet) -> Void) {
+        let traversalSuper = cast(superlayout)
+        let traversalSub = cast(sublayout)
+        traversalSuper?.traverse(superview, viewInfoSet: viewInfoSet, constraintHndler: { superview, subview, constraints, viewInfoSet in
+            handler(superview, subview, constraints, viewInfoSet)
+            traversalSub?.traverse(subview, viewInfoSet: viewInfoSet, constraintHndler: handler)
+        })
+    }
+}
+
+extension TupleLayout: LayoutTraversal {
+    
+    var traversals: [LayoutTraversal] {
+        Mirror(reflecting: layout).children.compactMap { (_, value) in
```

Comment:
wow! There was even a way to use the mirror like this!

Reply:
> wow! There was even a way to use the mirror like this!

its amazing to me too.